renku/core/commands/run.py

```diff
@@ -0,0 +1,155 @@
+    except OSError:
+        tty_exists = False
+
+    try:
+        stdout_redirected = "stdout" in mapped_std
+        stderr_redirected = "stderr" in mapped_std
+
+        if tty_exists:
+            # if renku was called with redirected stdout/stderr, undo the
+            # redirection here so error messages can be printed normally
+            if stdout_redirected:
+                system_stdout = open("/dev/tty", "w")
+                old_stdout = sys.stdout
+                sys.stdout = system_stdout
+
+            if stderr_redirected:
+                system_stderr = open("/dev/tty", "w")
+                old_stderr = sys.stderr
+                sys.stderr = system_stderr
+
+        working_dir = client.repo.working_dir
+        factory = CommandLineToolFactory(
+            command_line=command_line,
+            explicit_inputs=explicit_inputs,
+            explicit_outputs=explicit_outputs,
+            directory=os.getcwd(),
+            working_dir=working_dir,
+            no_input_detection=no_input_detection,
+            no_output_detection=no_output_detection,
+            successCodes=success_codes,
+            **{name: os.path.relpath(path, working_dir) for name, path in mapped_std.items()},
+        )
+        with factory.watch(client, no_output=no_output) as tool:
+            # Don't compute paths if storage is disabled.
+            if client.check_external_storage():
+                # Make sure all inputs are pulled from a storage.
+                paths_ = (path for _, path in tool.iter_input_files(client.workflow_path))
+                client.pull_paths_from_storage(*paths_)
+
+            if tty_exists:
+                # apply original output redirection
+                if stdout_redirected:
+                    sys.stdout = old_stdout
+                if stderr_redirected:
+                    sys.stderr = old_stderr
+
+            return_code = call(
+                factory.command_line, cwd=os.getcwd(), **{key: getattr(sys, key) for key in mapped_std.keys()},
+            )
+
+            sys.stdout.flush()
+            sys.stderr.flush()
+
+            if tty_exists:
+                # change back to /dev/tty redirection
+                if stdout_redirected:
+                    sys.stdout = system_stdout
+                if stderr_redirected:
+                    sys.stderr = system_stderr
+
+            if return_code not in (success_codes or {0}):
+                raise errors.InvalidSuccessCode(return_code, success_codes=success_codes)
+
+        client.process_and_store_run(command_line_tool=tool, name=name, client=client)
+
+        if factory.messages:
+            communication.echo(factory.messages)
+
+        if factory.warnings:
+            communication.echo(factory.warnings)
+
+    finally:
+        if system_stdout:
+            sys.stdout = old_stdout
+            system_stdout.close()
+        if system_stderr:
+            sys.stderr = old_stderr
+            system_stderr.close()
```

Comment:
```suggestion
        if system_stdout:
            system_stdout.flush()
            sys.stdout = old_stdout
            system_stdout.close()
        if system_stderr:
            system_stderr.flush()
            sys.stderr = old_stderr
            system_stderr.close()
```

Not really related to this PR, but I noticed sometimes renku errors will still get written to file instead of printed (in `renku run echo 'test' > test.txt` you could end up with a renku error in test.txt instead of printed in the terminal).

I think this could be solved by flushing `system_stdout` and `system_stderr` here, so would be nice if we could add that.